src/librustc/traits/error_reporting.rs

```diff
@@ -357,14 +361,20 @@ pub fn report_selection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
                             let trait_ref = trait_predicate.to_poly_trait_ref();
                             let mut err = struct_span_err!(
                                 infcx.tcx.sess, obligation.cause.span, E0277,
-                                "the trait `{}` is not implemented for the type `{}`",
-                                trait_ref, trait_ref.self_ty());
-
-                            // Check if it has a custom "#[rustc_on_unimplemented]"
-                            // error message, report with that message if it does
-                            let custom_note = report_on_unimplemented(infcx, &trait_ref.0,
-                                                                      obligation.cause.span);
-                            if let Some(s) = custom_note {
+                                "the predicate `{}` is not satisfied",
```

Comment:
Hmm, I wonder if it is better to write in this uniform style, or to try to have a custom error message? Not clear to me. It seems likely that if we are going to make this message generic, we could combine with some other error-code -- though maybe there is value in separating out this form of predicate from the others. NM, keep it with its own code.